retworkx-core/src/connectivity/conn_components.rs

```diff
@@ -0,0 +1,129 @@
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations
+// under the License.
+
+use hashbrown::HashSet;
+use std::collections::VecDeque;
+use std::hash::Hash;
+
+use petgraph::{
+    visit::{
+        EdgeCount, GraphProp, IntoEdges, IntoNodeIdentifiers, IntoNodeReferences, NodeIndexable,
+        VisitMap, Visitable,
+    },
+    Undirected,
+};
+
+/// Given an undirected graph, a start node and the visit_map for
+/// the graph, this function returns a connected component set.
+///
+/// :param Graph graph: The input graph to find the connected
+///     components for.
+/// :param NodeIndex start: The node to start from.
+/// :param Visitable::Map discovered: The visit map for the graph.
+///
+/// :return: A set of connected components for the start node
+/// :rtype: HashSet<usize>
+pub fn bfs_undirected<G>(graph: G, start: G::NodeId, discovered: &mut G::Map) -> HashSet<G::NodeId>
```

Comment:
> With current trait restrictions we can't really remove retworkx's crate version of these functions since they also work for _directed_ `StableGraph`s.

Can you elaborate on this for me? What are the trait restrictions? Thanks.

Reply:
I had the same thought. It's more of a helper function than a public one.

Reply:
Done in https://github.com/Qiskit/retworkx/pull/595/commits/17bb73a729c3845bea0f413d2344f2a6c6a7458b.